tabbycat/notifications/utils.py

```diff
@@ -171,3 +171,31 @@ def _create_motion_list():
             emails.append((context, speaker))
 
     return emails
+
+
+def team_speaker_email_generator(tournament_id):
+    emails = []
+    tournament = Tournament.objects.get(id=tournament_id)
+
+    for team in tournament.team_set.all().prefetch_related('speaker_set', 'break_categories').select_related('division', 'institution'):
+        context = {
+            'TOURN': str(tournament),
+            'SHORT': team.short_name,
+            'LONG': team.long_name,
+            'CODE': team.code_name,
+            'DIVISION': team.division.name,
+            'BREAK': _(", ").join([breakq.name for breakq in team.break_categories.all()]),
+            'SPEAKERS': _(", ").join([p.name for p in team.speaker_set.all()]),
+            'INSTITUTION': str(team.institution),
+            'EMOJI': team.emoji
+        }
+
+        for speaker in team.speakers:
+            if speaker.email is None:
+                continue
+
+            context['USER'] = speaker.name
+
+            emails.append((context, speaker))
```

Comment:
This line adds a reference to the dict `context`, not a copy (in Python, dicts are always passed by reference). Because you're modifying the same `context` once for each speaker, the `emails` list will have the last speaker in each team, multiple times, because you have the same dict for each speaker in a team.

The dict is different for different teams, because line 181 assigns the whole dictionary. It's just the same for all speakers on a single team.

The method `context.copy()` might help.

Reply:
Wouldn't I have seen that it doesn't work with the other email generator functions that use the same pattern (like `motion_release_email_generator` or `standings_email_generator`)?

Reply:
When I call `standings_email_generator` from the shell, this is exactly what happens:
``` python
In [2]: standings_email_generator("http://helloworld.org/", 25)
Out[2]:
[({'TOURN': 'emailstest',
   'URL': 'http://helloworld.org/',
   'POINTS': '0',
   'TEAM': 'Team Bravo',
   'USER': 'Andreas Grunewald'},
  <Speaker: Zhu Tsao>),
 ({'TOURN': 'emailstest',
   'URL': 'http://helloworld.org/',
   'POINTS': '0',
   'TEAM': 'Team Bravo',
   'USER': 'Andreas Grunewald'},
  <Speaker: Deodata Trevisani>),
 ({'TOURN': 'emailstest',
   'URL': 'http://helloworld.org/',
   'POINTS': '0',
   'TEAM': 'Team Bravo',
   'USER': 'Andreas Grunewald'},
  <Speaker: Frida Padovesi>),
 ({'TOURN': 'emailstest',
   'URL': 'http://helloworld.org/',
   'POINTS': '0',
   'TEAM': 'Team Bravo',
   'USER': 'Andreas Grunewald'},
  <Speaker: Andreas Grunewald>)]
```

I don't seem to be able to get emails working at all, though—I think something's broken for me with the new channels system. So I couldn't tell you what's in the emails. (They don't get logged, either, so it's like the consumer's not picking up on the message.)

Reply:
Aie! Guess I have more to fix then! Also, make sure a Redis server is running, and redis is enabled. Then the emails should work (other than this one apparently)